### ISSUE TITLE
[next/link] Using URL object with dynamic routes does not work

### ISSUE THREAD

**Issue by mantysalo** [vercel/next.js#17380]
# Bug report

## Describe the bug

The documentation for `next/link` describes using an URL object for the href prop:
https://nextjs.org/docs/api-reference/next/link#with-url-object

However, it seems that for dynamic routes, the value of the `slug` key from the query object just gets appended to the URL as a normal query parameter and the link ends up pointing to: `/blog/[slug]?slug=my-post`

This happens in both dev and production builds.

## To Reproduce

1. Go to https://github.com/mantysalo/next-link-issue
2. Clone the repo
3. Start dev server
4. Observe the URL of the link
5. See error

## Expected behavior

With an URL object as such:

```javascript
{
          pathname: '/blog/[slug]',
          query: { slug: 'my-post' },
}
```

I would expect the url to be: `/blog/my-post` as it is stated in the documentation.

## System information

- OS: macOS
- Version of Next.js: 9.5.3
- Version of Node.js: 12.16.0

**Comment by lfades** [comment]
@mantysalo Our docs are currently pointing to a version that includes a change from the canary release and is not yet out 😅 

A new stable will come out soon and the URL object should work as expected.

**Comment by CarsonMcKinstry** [comment]
@lfades Do you happen to know a rough timeline on the canary release? We ran into the same issue with our current app :)

**Comment by ijjk** [comment]
This is already available in the latest canary `v9.5.4-canary.23` so if you would like to take advantage of it you could update to that version or wait on the next stable release as @lfades mentioned which isn't too far out.

**Comment by mantysalo** [comment]
Thanks for the responses 👍 I'll wait for the next release.

**Comment by pixelass** [comment]
Why has this been closed with out being resolved? We have the same behavior using 

`router.push(...)`

Should I open a new issue or can we keep this one? @timneutkens

**Comment by lfades** [comment]
@pixelass This shouldn't be an issue anymore, make sure to be on the latest Next.js version, and if there is a bug then please open a new issue with a reproduction 🙏

**Comment by oktaykose** [comment]
`    const basePath = Router.asPath.split('?')[0] || Router.asPath
     
        const query =  Router.query
        query.lang = lang || locale

        Object.keys(query).map(function(k) {
          if(Router.pathname.includes(`[${k}]`)){
              delete query[k]
          }
        })

        const url = { pathname: Router.pathname, query }
        const urlAs = { pathname: basePath, query }
        
        Router.push(url, urlAs)`

**Comment by oktaykose** [comment]
`    const basePath = Router.asPath.split('?')[0] || Router.asPath
     
        const query =  Router.query
        query.lang = lang || locale

        Object.keys(query).map(function(k) {
          if(Router.pathname.includes(`[${k}]`)){
              delete query[k]
          }
        })

        const url = { pathname: Router.pathname, query }
        const urlAs = { pathname: basePath, query }
        
        Router.push(url, urlAs)`

**Comment by oktaykose** [comment]
const basePath = Router.asPath.split('?')[0] || Router.asPath
     
        const query =  Router.query
        query.lang = lang || locale

        Object.keys(query).map(function(k) {
          if(Router.pathname.includes(`[${k}]`)){
              delete query[k]
          }
        })

        const url = { pathname: Router.pathname, query }
        const urlAs = { pathname: basePath, query }
        
        Router.push(url, urlAs)

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
